File: AlbumLibrary/src/main/java/com/album/ui/activity/AlbumActivity.java
package com.album.ui.activity;

import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.ListPopupWindow;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.RelativeLayout;

import com.album.Album;
import com.album.AlbumConstant;
import com.album.R;
import com.album.entity.FinderEntity;
import com.album.ui.adapter.ListPopupWindowAdapter;
import com.album.annotation.PermissionsType;
import com.album.ui.fragment.AlbumFragment;
import com.album.ui.view.AlbumMethodActivityView;
import com.album.util.AlbumTool;

import java.util.List;
import java.util.Objects;

/**
 * by y on 14/08/2017.
 */

public class AlbumActivity extends AlbumBaseActivity
        implements View.OnClickListener, AlbumMethodActivityView, AdapterView.OnItemClickListener {

    private Toolbar toolbar;
    private AppCompatTextView preview;
    private AppCompatTextView select;
    private AlbumFragment albumFragment;
    private AppCompatTextView finderTv;
    private ListPopupWindow listPopupWindow;
    private RelativeLayout albumBottomView;

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (albumFragment != null) {
            albumFragment.disconnectMediaScanner();
            albumFragment = null;
        }
        if (listPopupWindow != null) {
            listPopupWindow = null;
        }
    }

    @Override
    protected void initCreate(@Nullable Bundle savedInstanceState) {
        if (albumFragment != null) {
            albumFragment = null;
        }
        initFragment();
        initBottomView();
        initListPopupWindow();
    }

    @Override
    protected void initView() {
        toolbar = findViewById(R.id.album_toolbar);
        preview = findViewById(R.id.album_tv_preview);
        select = findViewById(R.id.album_tv_select);
        finderTv = findViewById(R.id.album_tv_finder_all);
        albumBottomView = findViewById(R.id.album_bottom_view);
        listPopupWindow = new ListPopupWindow(this);
        preview.setOnClickListener(this);
        select.setOnClickListener(this);
        finderTv.setOnClickListener(this);
        preview.setVisibility(albumConfig.isRadio() ? View.GONE : View.VISIBLE);
        select.setVisibility(albumConfig.isRadio() ? View.GONE : View.VISIBLE);
    }

    @Override
    public void initFragment() {
        FragmentManager supportFragmentManager = getSupportFragmentManager();
        Fragment fragment = supportFragmentManager.findFragmentByTag(AlbumFragment.class.getSimpleName());
        if (fragment != null) {
            albumFragment = (AlbumFragment) fragment;
        } else {
            albumFragment = AlbumFragment.newInstance();
        }
        supportFragmentManager
                .beginTransaction()
                .replace(R.id.album_frame, albumFragment, AlbumFragment.class.getSimpleName())
                .commit();
    }

    @Override
    public void initBottomView() {
        finderTv.setText(TextUtils.isEmpty(finderName) ? getString(R.string.album_all) : finderName);
        albumBottomView.setBackgroundColor(ContextCompat.getColor(this, albumConfig.getAlbumBottomViewBackground()));
        finderTv.setTextSize(albumConfig.getAlbumBottomFinderTextSize());
        finderTv.setTextColor(ContextCompat.getColor(this, albumConfig.getAlbumBottomFinderTextColor()));
        finderTv.setCompoundDrawables(null, null, AlbumTool.getDrawable(this, albumConfig.getAlbumBottomFinderTextCompoundDrawable(), albumConfig.getAlbumBottomFinderTextDrawableColor()), null);
        if (albumConfig.getAlbumBottomFinderTextBackground() != -1) {
            finderTv.setBackgroundResource(albumConfig.getAlbumBottomFinderTextBackground());
        }

        preview.setText(albumConfig.getAlbumBottomPreViewText());
        preview.setTextSize(albumConfig.getAlbumBottomPreViewTextSize());
        preview.setTextColor(ContextCompat.getColor(this, albumConfig.getAlbumBottomPreViewTextColor()));
        if (albumConfig.getAlbumBottomPreviewTextBackground() != -1) {
            preview.setBackgroundResource(albumConfig.getAlbumBottomPreviewTextBackground());
        }

        select.setText(albumConfig.getAlbumBottomSelectText());
        select.setTextSize(albumConfig.getAlbumBottomSelectTextSize());
        select.setTextColor(ContextCompat.getColor(this, albumConfig.getAlbumBottomSelectTextColor()));
        if (albumConfig.getAlbumBottomSelectTextBackground() != -1) {
            select.setBackgroundResource(albumConfig.getAlbumBottomSelectTextBackground());
        }
    }

    @Override
    public void initListPopupWindow() {
        listPopupWindow.setAnchorView(finderTv);
        listPopupWindow.setWidth(albumConfig.getAlbumListPopupWidth());
        listPopupWindow.setHorizontalOffset(albumConfig.getAlbumListPopupHorizontalOffset());
        listPopupWindow.setVerticalOffset(albumConfig.getAlbumListPopupVerticalOffset());
        listPopupWindow.setModal(true);
        listPopupWindow.setOnItemClickListener(this);
    }


    @Override
    protected void initTitle() {
        toolbar.setTitle(albumConfig.getAlbumToolbarText());
        toolbar.setTitleTextColor(ContextCompat.getColor(this, albumConfig.getAlbumToolbarTextColor()));
        Drawable drawable = ContextCompat.getDrawable(this, albumConfig.getAlbumToolbarIcon());
        Objects.requireNonNull(drawable).setColorFilter(ContextCompat.getColor(this, albumConfig.getAlbumToolbarIconColor()), PorterDuff.Mode.SRC_ATOP);
        toolbar.setNavigationIcon(drawable);
        toolbar.setBackgroundColor(ContextCompat.getColor(this, albumConfig.getAlbumToolbarBackground()));
        if (AlbumTool.hasL()) {
            toolbar.setElevation(albumConfig.getAlbumToolbarElevation());
        }
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Album.getInstance().getAlbumListener().onAlbumActivityFinish();
                finish();
            }
        });
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_album;
    }

    @Override
    protected void permissionsDenied(@PermissionsType int type) {
        Album.getInstance().getAlbumListener().onAlbumPermissionsDenied(type);
        if (albumConfig.isPermissionsDeniedFinish()) {
            finish();
        }
    }

    @Override
    protected void permissionsGranted(int type) {
        switch (type) {
            case AlbumConstant.TYPE_PERMISSIONS_ALBUM:
                albumFragment.onScanAlbum(null, false, false);
                break;
            case AlbumConstant.TYPE_PERMISSIONS_CAMERA:
                albumFragment.openCamera();
                break;
        }
    }

    @Override
    public void onClick(View v) {
        if (albumFragment == null) {
            Album.getInstance().getAlbumListener().onAlbumFragmentNull();
            return;
        }
        int i = v.getId();
        if (i == R.id.album_tv_preview) {
            albumFragment.multiplePreview();
        } else if (i == R.id.album_tv_select) {
            albumFragment.multipleSelect();
        } else if (i == R.id.album_tv_finder_all) {
            List<FinderEntity> finderEntity = albumFragment.getFinderEntity();
            if (finderEntity != null && !finderEntity.isEmpty()) {
                listPopupWindow.setAdapter(new ListPopupWindowAdapter(finderEntity));
                listPopupWindow.show();
                ListView listView = listPopupWindow.getListView();
                if (listView != null) {
                    listView.setBackgroundColor(ContextCompat.getColor(this, albumConfig.getAlbumListPopupItemBackground()));
                }
                return;
            }
            Album.getInstance().getAlbumListener().onAlbumFinderNull();
        }
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        ListView listView = listPopupWindow.getListView();
        if (listView == null) {
            return;
        }
        ListPopupWindowAdapter adapter = (ListPopupWindowAdapter) listView.getAdapter();
        FinderEntity finder = adapter.getFinder(position);
        if (finder == null) {
            return;
        }
        finderName = finder.getDirName();
        finderTv.setText(finder.getDirName());
        albumFragment.onScanAlbum(finder.getBucketId(), true, false);
        listPopupWindow.dismiss();
    }

    @Override
    public void onBackPressed() {
        Album.getInstance().getAlbumListener().onAlbumActivityBackPressed();
        super.onBackPressed();
    }
}

File: AlbumLibrary/src/main/java/com/album/ui/activity/AlbumActivity.kt
package com.album.ui.activity

import android.graphics.PorterDuff
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.widget.AppCompatTextView
import android.support.v7.widget.ListPopupWindow
import android.support.v7.widget.Toolbar
import android.text.TextUtils
import android.view.View
import android.widget.AdapterView
import android.widget.RelativeLayout
import com.album.Album
import com.album.R
import com.album.ui.adapter.FinderAdapter
import com.album.ui.fragment.AlbumFragment
import com.album.ui.view.AlbumFragmentToAtyListener
import com.album.util.AlbumTool

/**
 * by y on 14/08/2017.
 */

class AlbumActivity : AlbumBaseActivity(),
        View.OnClickListener,
        AlbumFragmentToAtyListener,
        AdapterView.OnItemClickListener {

    private lateinit var toolbar: Toolbar
    private lateinit var preview: AppCompatTextView
    private lateinit var select: AppCompatTextView
    private lateinit var finderTv: AppCompatTextView
    private lateinit var listPopupWindow: ListPopupWindow
    private lateinit var albumBottomView: RelativeLayout
    private lateinit var albumFragment: AlbumFragment

    override fun onDestroy() {
        albumFragment.disconnectMediaScanner()
        super.onDestroy()
    }

    override fun initCreate(savedInstanceState: Bundle?) {
        initFragment()
        initBottomView()
        initFinderView()
    }

    override fun initView() {
        toolbar = findViewById(R.id.album_toolbar)
        preview = findViewById(R.id.album_tv_preview)
        select = findViewById(R.id.album_tv_select)
        finderTv = findViewById(R.id.album_tv_finder_all)
        albumBottomView = findViewById(R.id.album_bottom_view)
        listPopupWindow = ListPopupWindow(this)
        preview.setOnClickListener(this)
        select.setOnClickListener(this)
        finderTv.setOnClickListener(this)
        preview.visibility = if (albumConfig.isRadio) View.GONE else View.VISIBLE
        select.visibility = if (albumConfig.isRadio) View.GONE else View.VISIBLE
    }

    private fun initFragment() {
        val supportFragmentManager = supportFragmentManager
        val fragment = supportFragmentManager.findFragmentByTag(AlbumFragment::class.java.simpleName)
        if (fragment != null) {
            albumFragment = fragment as AlbumFragment
            return
        }
        albumFragment = AlbumFragment.newInstance()
        supportFragmentManager
                .beginTransaction()
                .apply { add(R.id.album_frame, albumFragment, AlbumFragment::class.java.simpleName) }
                .commit()
    }

    private fun initBottomView() {
        finderTv.text = if (TextUtils.isEmpty(albumFragment.finderName)) getString(R.string.album_all) else albumFragment.finderName
        albumBottomView.setBackgroundColor(ContextCompat.getColor(this, albumConfig.albumBottomViewBackground))
        finderTv.textSize = albumConfig.albumBottomFinderTextSize.toFloat()
        finderTv.setTextColor(ContextCompat.getColor(this, albumConfig.albumBottomFinderTextColor))
        finderTv.setCompoundDrawables(null, null, AlbumTool.getDrawable(this, albumConfig.albumBottomFinderTextCompoundDrawable, albumConfig.albumBottomFinderTextDrawableColor), null)
        if (albumConfig.albumBottomFinderTextBackground != -1) {
            finderTv.setBackgroundResource(albumConfig.albumBottomFinderTextBackground)
        }
        preview.setText(albumConfig.albumBottomPreViewText)
        preview.textSize = albumConfig.albumBottomPreViewTextSize.toFloat()
        preview.setTextColor(ContextCompat.getColor(this, albumConfig.albumBottomPreViewTextColor))
        if (albumConfig.albumBottomPreviewTextBackground != -1) {
            preview.setBackgroundResource(albumConfig.albumBottomPreviewTextBackground)
        }
        select.setText(albumConfig.albumBottomSelectText)
        select.textSize = albumConfig.albumBottomSelectTextSize.toFloat()
        select.setTextColor(ContextCompat.getColor(this, albumConfig.albumBottomSelectTextColor))
        if (albumConfig.albumBottomSelectTextBackground != -1) {
            select.setBackgroundResource(albumConfig.albumBottomSelectTextBackground)
        }
    }

    private fun initFinderView() {
        listPopupWindow.anchorView = finderTv
        listPopupWindow.width = albumConfig.albumListPopupWidth
        listPopupWindow.horizontalOffset = albumConfig.albumListPopupHorizontalOffset
        listPopupWindow.verticalOffset = albumConfig.albumListPopupVerticalOffset
        listPopupWindow.isModal = true
        listPopupWindow.setOnItemClickListener(this)
    }


    override fun initTitle() {
        toolbar.setTitle(albumConfig.albumToolbarText)
        toolbar.setTitleTextColor(ContextCompat.getColor(this, albumConfig.albumToolbarTextColor))
        val drawable = ContextCompat.getDrawable(this, albumConfig.albumToolbarIcon)
        drawable?.setColorFilter(ContextCompat.getColor(this, albumConfig.albumToolbarIconColor), PorterDuff.Mode.SRC_ATOP)
        toolbar.navigationIcon = drawable
        toolbar.setBackgroundColor(ContextCompat.getColor(this, albumConfig.albumToolbarBackground))
        if (AlbumTool.hasL()) {
            toolbar.elevation = albumConfig.albumToolbarElevation
        }
        toolbar.setNavigationOnClickListener {
            Album.instance.albumListener.onAlbumActivityFinish()
            finish()
        }
    }

    override fun getLayoutId(): Int = R.layout.album_activity_album

    override fun onClick(v: View) {
        when (v.id) {
            R.id.album_tv_preview -> albumFragment.multiplePreview()
            R.id.album_tv_select -> albumFragment.multipleSelect()
            R.id.album_tv_finder_all -> {
                val finderEntity = albumFragment.getFinderEntity()
                if (!finderEntity.isEmpty()) {
                    listPopupWindow.setAdapter(FinderAdapter(finderEntity))
                    listPopupWindow.show()
                    val listView = listPopupWindow.listView
                    listView?.setBackgroundColor(ContextCompat.getColor(this, albumConfig.albumListPopupItemBackground))
                    return
                }
                Album.instance.albumListener.onAlbumFinderNull()
            }
        }
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val listView = listPopupWindow.listView ?: return
        val adapter = listView.adapter as FinderAdapter
        val finder = adapter.getFinder(position)
        albumFragment.finderName = finder.dirName
        finderTv.text = finder.dirName
        albumFragment.onScanAlbum(finder.bucketId, true, false)
        listPopupWindow.dismiss()
    }

    override fun onChangedFinderName(name: String) {
        finderTv.text = name
    }

    override fun onBackPressed() {
        Album.instance.albumListener.onAlbumActivityBackPressed()
        super.onBackPressed()
    }
}
